include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -735,179 +734,307 @@ void instr_remainder_and_pdf_execute(in instr_t instr, in irr_glsl_BSDFSample s)
-			pdf = irr_glsl_beckmann_pdf(s, currInteraction.isotropic, a2);
-#else
-			bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_smith_height_correlated_cos_eval_DG(currBSDFParams, currInteraction, a, a2, ay, ay2);
-			pdf = irr_glsl_beckmann_pdf(s, currInteraction, a, a2, ay, ay2);
+		float G1_over_2NdotV = 0.0;
+		float G2_over_G1 = 0.0;
+		float ndf_val = 0.0;
+
+		BEGIN_CASES(ndf)
+#ifdef NDF_GGX
+		CASE_BEGIN(ndf, NDF_GGX) {
+			G1_over_2NdotV = irr_glsl_GGXSmith_G1_wo_numerator(NdotV, TdotV2, BdotV2, NdotV2, a2, ay2);
+			G2_over_G1 = irr_glsl_ggx_smith_G2_over_G1(NdotL, s.TdotL*s.TdotL, s.BdotL*s.BdotL, s.NdotL*s.NdotL, NdotV, TdotV2, BdotV2, NdotV2, a2, ay2);
+			ndf_val = irr_glsl_ggx_aniso(TdotH2, BdotH2, NdotH2, a, ay, a2, ay2);
+		} CASE_END
 #endif
 
-		} else
+#ifdef NDF_BECKMANN
+		CASE_BEGIN(ndf, NDF_BECKMANN) {
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, a2, ay2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL*s.TdotL, s.BdotL*s.BdotL, s.NdotL*s.NdotL, a2, ay2);
+			ndf_val = irr_glsl_beckmann(ax, ay, a2, ay2, TdotH2, BdotH2, NdotH2);
+		} CASE_END
 #endif
 
 #ifdef NDF_PHONG
-		if (ndf==NDF_PHONG) {
-
-#ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, n, a2);
-			pdf = irr_glsl_beckmann_pdf(s, currInteraction.isotropic, a2);
-#else
-			bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(currBSDFParams, currInteraction, n, ny, a2, ay2);
-			pdf = irr_glsl_beckmann_pdf(s, currInteraction, a, a2, ay, ay2);
+		CASE_BEGIN(ndf, NDF_PHONG) {
+			float lambdaV = irr_glsl_smith_beckmann_Lambda(TdotV2, BdotV2, NdotV2, ax2, ay2);
+			G1_over_2NdotV = irr_glsl_smith_G1(lambdaV) / (2.0 * NdotV);
+			G2_over_G1 = irr_glsl_beckmann_smith_G2_over_G1(lambdaV + 1.0, s.TdotL*s.TdotL, s.BdotL*s.BdotL, s.NdotL*s.NdotL, ax2, ay2);
+			ndf_val = irr_glsl_beckmann(ax, ay, ax2, ay2, TdotH2, BdotH2, NdotH2);
+		} CASE_END
 #endif
+		CASE_OTHERWISE
+		{} //else "empty braces"
+		END_CASES
 
-		} else
-#endif
+		bxdf_eval_scalar_part = 0.5 * G2_over_G1 * G1_over_2NdotV * ndf_val; //cos*D*G2/(4*cos*cos)
+		pdf = irr_glsl_smith_VNDF_pdf_wo_clamps(ndf_val, G1_over_2NdotV);
 
-#ifndef ONLY_ONE_NDF
-		{} //else "empty braces"
+		float VdotH = microfacet.isotropic.VdotH;
+		float VdotH_clamp = irr_glsl_conditionalAbsOrMax(is_bsdf, VdotH, 0.0);
+		vec3 fr;
+#ifdef OP_CONDUCTOR
+		if (op == OP_CONDUCTOR)
+			fr = irr_glsl_fresnel_conductor(ior[0], ior[1], VdotH_clamp);
+		else
 #endif
-	}
+			fr = irr_glsl_fresnel_dielectric_common(ior2[0], VdotH_clamp);
 
-	uvec3 regs = instr_decodeRegisters(instr);
-#ifdef OP_DIFFUSE
-	if (op==OP_DIFFUSE) {
-		instr_execute_cos_eval_DIFFUSE(instr, regs, params, bsdf_data);
-	} else
+		if (is_bsdf && is_valid)
```

Comment:
this check should move up... a lot (if bsdf and not valid, you can skip all this)

Reply:
basically it should be
```glsl
float pdf = 0.0;
vec3 remainder = vec3(0.0);
if (cosFactor>FLT_MIN)
{
    bool is_valid = true;

    if (is_bsdf)
    {
        is_valid = //special check
    }

    if (is_valid)
    {
         //the meat of the function
    }
}
writeRegisters();
```

Reply:
I messed up and gave bad advice